src/dialogs/make-dialog-manager.ts

```diff
@@ -60,11 +72,46 @@ export const showDialog = async (
       }
       return;
     }
-    LOADED[dialogTag] = dialogImport().then(() => {
-      const dialogEl = document.createElement(dialogTag) as HassDialog;
-      element.provideHass(dialogEl);
-      return dialogEl;
-    });
+    LOADED[dialogTag] = {
+      element: dialogImport().then(() => {
+        const dialogEl = document.createElement(dialogTag) as HassDialog;
+        element.provideHass(dialogEl);
+        return dialogEl;
+      }),
+      closedFocusTarget: null,
+      closedFocusDialog: null,
+    };
+  }
+
+  // Go down all components to find the active element,
+  // but keep the original focus target if dialog is being replaced
+  if (mainWindow.history.state?.replaced) {
+    LOADED[dialogTag].closedFocusTarget =
+      LOADED[mainWindow.history.state.dialog].closedFocusTarget;
+    LOADED[dialogTag].closedFocusDialog =
+      LOADED[mainWindow.history.state.dialog].closedFocusDialog;
+  } else {
+    let focusedDialog: LoadedDialogInfo["closedFocusDialog"] = null;
+    let focusedElement = document.activeElement;
+    while (focusedElement?.shadowRoot?.activeElement) {
+      focusedElement = focusedElement.shadowRoot.activeElement;
+
+      // Detect when focus is slotted into certain components
+      const slotRoot = focusedElement.assignedSlot?.getRootNode();
+      if (slotRoot instanceof ShadowRoot) {
+        const slotHost = slotRoot.host;
+        if (slotHost instanceof HaButtonMenu) {
+          // Use trigger button since popup menu will be hidden again
+          focusedElement = slotHost.querySelector('[slot="trigger"]');
+          break;
+        } else if (slotHost instanceof HaDialog) {
+          // Focus will return inside another dialog
+          focusedDialog = slotHost;
+        }
```

Comment:
Can we instead add a function to the elements that need this, check if that function exists and then let that function return the correct element? So the logic is in the element and not in the dialog manager?

Reply:
My initial reaction was yes, but after thinking about how to code it, it won't really remove the logic from here because what to do with the returned value depends on the type of element.  This module would still need to know whether it's modifying the focus target or just recording an underlay dialog.

If you just want this logic out of the dialog module, I think I have a much better idea.  Give me a day or two to add a commit.

Reply:
I moved all of the active element/dialog searching to the `deepActiveElement` function, which now has 3 different modes.  So all of that is now in one place.  The rest of the code left in the dialog manager just grabs it, stores it, and sets the event handler.  See if that resolves your concerns.

Apologies for the delay... had to learn some unfamiliar Typescript then work around a known limitation.